eth/eventhandler/handlers.go

```diff
@@ -469,6 +469,48 @@ func (eh *EventHandler) handleFeeRecipientAddressUpdated(txn basedb.Txn, event *
 	return r != nil, nil
 }
 
+type ExitDescriptor struct {
+	PubKey         phase0.BLSPubKey
+	ValidatorIndex phase0.ValidatorIndex
+	Slot           phase0.Slot
+}
+
+func (eh *EventHandler) handleValidatorExited(txn basedb.Txn, event *contract.ContractValidatorExited) (*ExitDescriptor, error) {
+	logger := eh.logger.With(
+		fields.EventName(ValidatorExited),
+		fields.TxHash(event.Raw.TxHash),
+		fields.PubKey(event.PublicKey.Bytes()), // TODO: check if logged correctly
+		fields.OperatorIDs(event.OperatorIds),
+	)
+	logger.Debug("processing event")
+	defer logger.Debug("processed event")
+
+	share := eh.nodeStorage.Shares().Get(txn, event.PublicKey.Bytes())
+	if share == nil {
+		logger.Warn("malformed event: could not find validator share")
+		return nil, &MalformedEventError{Err: ErrValidatorShareNotFound}
+	}
+
+	if !share.BelongsToOperator(eh.operatorData.GetOperatorData().ID) {
+		return nil, nil
+	}
+
+	if share.BeaconMetadata == nil {
+		return nil, nil
+	}
+
+	pk := phase0.BLSPubKey{}
+	copy(pk[:], share.ValidatorPubKey)
+
+	ed := &ExitDescriptor{
+		PubKey:         pk,
+		ValidatorIndex: share.BeaconMetadata.Index,
+		Slot:           eh.networkConfig.Beacon.EstimatedCurrentSlot(),
```

Comment:
we need to see how this plays out, what if its the end of the slot? might be ok since it's not full consensus. Also we need to see if the duty slot should be the same as the signing slot, as again, until we finish and publish this slot might be over, so we might want to sign on exit for next slot even though we run the duty in the current slot.

Reply:
it seems that the beacon chain only cares that the slot and epoch of the exit request isn't in the far future, or too near to the validator activation.
see ethereum spec for more info -
```python
//    def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVoluntaryExit) -> None:
//     voluntary_exit = signed_voluntary_exit.message
//     validator = state.validators[voluntary_exit.validator_index]
//     # Verify the validator is active
//     assert is_active_validator(validator, get_current_epoch(state))
//     # Verify exit has not been initiated
//     assert validator.exit_epoch == FAR_FUTURE_EPOCH
//     # Exits must specify an epoch when they become valid; they are not valid before then
//     assert get_current_epoch(state) >= voluntary_exit.epoch
//     # Verify the validator has been active long enough
//     assert get_current_epoch(state) >= validator.activation_epoch + SHARD_COMMITTEE_PERIOD
//     # Verify signature
//     domain = get_domain(state, DOMAIN_VOLUNTARY_EXIT, voluntary_exit.epoch)
//     signing_root = compute_signing_root(voluntary_exit, domain)
//     assert bls.Verify(validator.pubkey, signing_root, signed_voluntary_exit.signature)
//     # Initiate exit
```

Reply:
@y0sher @nkryuchkov 
the scheduling is indeed problematic, as different operators may handle this event at different slots, which may happen surprisingly often when your EL/CL node is just a few seconds behind.

in case 2/4 operators schedule at different slots, they wouldn't reach consensus and force the user to pay fees on another TX

i believe the most reliable way to schedule this duty must be deterministic (without current time, determined from the event only), one proposal could be:
1. when handling the event, pass over the event's block number (rather than current slot)
2. when executing the task, determine the event's slot (requires querying the EL) and schedule the duty for X slots ahead

this means that if X is 4 (for example), then operators can handle the event up to 3 slots late and still execute it together and reach consensus

wdyt? do you have any other ideas that don't require querying the EL?

Reply:
so again this sums up the first part I mentioned where the issue is with scheduling the slot when we do the duty. which we need to somehow be able to determine same slot. using the `CurrentEstimatedSlot` probably won't fit since @moshe-blox says BL/EL nodes might delay and the slots end up different. (even though so far on stage this never happened). 
So I think we must hang to something deterministic about the event to be able to determine a slot. it seems like everything will require contact with EL. we should get this finalized ASAP though, @nkryuchkov ,@lior-blox, @GalRogozinski do you have any idea or leaning towards any approach?